fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/ExplainTokens.java

```diff
@@ -0,0 +1,1110 @@
+            Verify.verify(tokenSizes.length == minLengths.length);
+            return tokenSizes.length - 1;
+        }
+
+        protected boolean isRenderingEnabled(final int renderingExplainLevel) {
+            return getExplainLevel() >= renderingExplainLevel;
+        }
+
+        @SuppressWarnings("unchecked")
+        protected <T extends Token> CharSequence renderIfEnabled(final int renderingExplainLevel,
+                                                                 @Nonnull final String stringedToken,
+                                                                 @Nonnull final BiFunction<T, CharSequence, CharSequence> renderFunction) {
+            if (isRenderingEnabled(renderingExplainLevel)) {
+                return renderFunction.apply((T)this, stringedToken);
+            }
+            return "";
+        }
+
+        protected CharSequence cutOffIfNeeded(final int remainingCharacterBudget,
+                                              @Nonnull final CharSequence stringedToken) {
+            Verify.verify(remainingCharacterBudget >= 0);
+            if (remainingCharacterBudget >= stringedToken.length()) {
+                return stringedToken;
+            }
+            return new StringBuilder(stringedToken.subSequence(0, remainingCharacterBudget)).append("...");
+        }
+
+
+        public int getTokenSize(final int explainLevel) {
+            return explainLevel < tokenSizes.length ? tokenSizes[explainLevel] : 0;
+        }
+
+        public int getMinLength(final int explainLevel) {
+            return explainLevel < minLengths.length ? minLengths[explainLevel] : 0;
+        }
+
+        public int getMaxLength(final int explainLevel) {
+            return explainLevel < maxLengths.length ? maxLengths[explainLevel] : 0;
+        }
+
+        @Nonnull
+        public abstract CharSequence render(int renderingExplainLevel,
+                                            @Nonnull ExplainFormatter explainFormatter,
+                                            int remainingCharacterBudget);
+
+        @Nonnull
+        private static int[] vectorForExplainLevel(final int explainLevelExclusive, int value) {
+            final int[] result = new int[explainLevelExclusive];
+            Arrays.fill(result, value);
+            return result;
+        }
+    }
+
+    /**
+     * A push token. This kind of token is used to interact with the symbol table of the formatter and does not render
+     * into anything.
+     */
+    public static class PushToken extends Token {
+        public PushToken(final int explainLevel) {
+            super(TokenKind.PUSH, explainLevel, 0, 0);
+        }
+
+        @Nonnull
+        @Override
+        public String render(final int renderingExplainLevel, @Nonnull final ExplainFormatter explainFormatter, final int remainingCharacterBudget) {
+            if (isRenderingEnabled(renderingExplainLevel)) {
+                explainFormatter.pushScope();
+            }
```

Comment:
It's interesting that the scope push/pop is skipped at certain rendering levels. Is the idea that if the rendering level is too coarse, then everything referencing the scope will not be rendered? Would it be okay to just always push and pop the scope, regardless of the format level?

Reply:
it could work but the premise is that a token that is not rendered should also have no side-effects.